topdown/trusted_identity_providers.go

```diff
@@ -0,0 +1,121 @@
+package topdown
+
+import (
+	"errors"
+	"fmt"
+	"github.com/coreos/go-oidc"
+	jwt "github.com/square/go-jose/v3/jwt"
+	"golang.org/x/net/context"
+	"sync"
+	"time"
+)
+
+
+var globalTrustedIdProviderManager TrustedIdProviderManagerImpl
+var globalTrustedIssuerContextTimeout = time.Second * time.Duration(5)
+
+// May be used to manage an token verification against an entire collection
+// of trusted IdP's.
+type TrustedIdProviderManager interface {
+
+	// On nil error, IDToken contains the parsed jwt token iff there exist
+	// at least one trusted identity provider in this managed set of IdP's
+	// that successfully validated this token. Else, (if not valid or not
+	// trusted or an error is encountered), a non-nil error returned and
+	// IDToken is nil.
+	VerifyToken(token *string) (*oidc.IDToken, error)
+}
+
+// Private struct to implement interface, TrustedIdProviderManager.
+// We use a sync map to safely manage a collection of trusted
+// issuers and their verifiers.
+type TrustedIdProviderManagerImpl struct {
+
+	// Thread safe Map of string of trustedIssuer to *oidc.IDTokenVerifier trustedIssuerVerifier.
+	trustedVerifiers sync.Map
+
+}
+
+func CreateOrGetVerifier(idp *string) (*oidc.IDTokenVerifier, error) {
+
+	// If we already have a verifier for this issuer, use it.
+	if loadedVerifier, ok := globalTrustedIdProviderManager.trustedVerifiers.Load(*idp); ok {
+		return loadedVerifier.(*oidc.IDTokenVerifier), nil
+	}
+
+	// Create a new issuer verifier, save it, and use it.
+	ctx, _ := context.WithTimeout(context.Background(), globalTrustedIssuerContextTimeout)
```

Comment:
We might want to expose the timeout to the caller of the builtin function since, as I understand it, this will go off and make a network request. Signature-wise we could maybe just add an options object or something as the third parameter, or maybe change the second parameter to be an object with one of the keys being for the list of IdP's?

Reply:
Ah, thanks for the comment.  You are correct, there are (possible) outbound network calls happening to fetch the IdP's public keys. About the user-facing option, it would be nice to not be forced to pick some reasonable timeout, but maybe there's something we can do here.

2 thoughts come to mind:

(1) This can't be the only outbound network call OPA is makes.  Sort of makes me wonder what the other timeout's might already be, and, maybe just borrow that timeout value? (to keep the interface simple)

(2) Maybe there's a config that should hold the details?

Reply:
For evaluation-time builtins its the second one. Only `http.send` does right now and it doesn't have a specified timeout .. and there is a bug open to fix that 😆 

I think the best option is to just provide an options object as the third parameter. While we could add something to the opa config file that is only specific to the OPA server/repl. Plus eventually someone is going to want it to be dynamic with the policy.. so we might as well just start there.